Assets/Scripts/Controllers/CharacterController.cs

```diff
@@ -23,6 +32,28 @@ private void FixedUpdate()
     }
 
 
+    private void Update()
+    {
+        // Check if the player has fired a weapon recently
+        if (_fireRate == 0)
+        {
+            if(Input.GetButton("Fire1"))
```

Comment:
This is the Legacy way of handling Unity Input response. Use of the new InputSystem is preferred. Fortunately in this case, it shouldn't actually be that difficult to refactor:

1. Move everything from Line 42 to 48 to a new method called `private void TryFire()`.
2. Create a new method `private void OnShoot(InputValue value) => TryFire();`. This event handler will call the TryFire method.
3. In the `CharacterMovement` InputActions file (Project window, double click to open), add a new Action called 'Shoot'.
4. Add a new binding for 'Shoot', set that binding to Left Mouse Button.
5. Everything should work the same as before, but without polling for input every frame. Let me know if this is not the case.